enzyme/Enzyme/MLIR/Passes/EnzymeBatchDiffPass.cpp

```diff
@@ -75,377 +316,798 @@ namespace {
-        auto iattr = cast<ActivityAttr>(ract);
-        auto val = iattr.getValue();
-        retActivity.push_back(val);
-      }
-      batchutils::BatchDiffCacheKey key{fnOp, in_args, inActivity, retActivity};
-
-      auto mergeItr = toMerge.find(key);
-      if (mergeItr != toMerge.end()) {
-        mergeItr->second.push_back(dop);
-      } else {
-        SmallVector<enzyme::ForwardDiffOp> v;
-        v.push_back(dop);
-        toMerge[key] = v;
+      for (auto fwdOp : blk->getOps<enzyme::ForwardDiffOp>()) {
+        auto fnOp = dyn_cast_or_null<FunctionOpInterface>(
+            symbolTable.lookupNearestSymbolFrom(fwdOp, fwdOp.getFnAttr()));
+        if (!fnOp)
+          continue;
+
+        batchutils::BatchDiffCacheKey key =
+            batchutils::createDiffCacheKey(fwdOp, fnOp);
+
+        toMerge[key].push_back(fwdOp);
       }
-      return mlir::WalkResult::advance();
-    });
 
-    OpBuilder builder(op);
+      for (auto &pair : toMerge) {
+        auto key = pair.first;
+        auto allDiffs = pair.second;
+        if (allDiffs.size() < 2)
+          continue;
+
+        // Collect inner effects of function
+        SmallVector<MemoryEffects::EffectInstance> calleeEffects =
+            batchutils::collectFnEffects(innerEffectCache, key.function);
+
+        // TODO skip if known readonly from existing analyses
+
+        bool skipMergeEntry = false;
+        // Map callee(primal function) memory effects to the calling
+        // function's(autodiff op's) memory effects. This allows us to also
+        // reason about memory effects on the derivative argument potentially
+        // being passed in(if the primal argument has activity enzyme_dup)
+
+        // ForwardDiff read(primal) ?-> read(deriv);
+        // write(primal) ?-> write(deriv)
+        //
+        // ReverseDiff read(primal) ?-> read or write(deriv);
+        // write(primal) ?-> read or write(deriv);
+        //
+        // Unknown effects are retained in the caller, and will always alias to
+        // true with any other effect
+
+        SmallVector<MemoryEffects::EffectInstance, 4> callerEffects;
+
+        for (auto &eff : calleeEffects) {
+          if (isa<MemoryEffects::Read>(eff.getEffect()) ||
+              isa<MemoryEffects::Write>(eff.getEffect())) {
+            if (Value resource = eff.getValue()) {
+
+              auto argnum = 0;
+              bool found = false;
+              for (auto fnArg : key.function.getArguments()) {
+                if (fnArg == resource) {
+                  argnum = fnArg.getArgNumber();
+                  found = true;
+                  break;
+                }
+              }
+              auto e = eff.getEffect();
+              if (found) {
+
+                // Primal value as effect
+                Value primalVal = key.inputs[argnum];
+                if (auto primalOpResult = dyn_cast<OpResult>(primalVal))
+                  callerEffects.emplace_back(MemoryEffects::EffectInstance(
```

Comment:
this is wrong, we need to be conservative and set this equal to the default resource